File: java/src/main/java/org/example/Game.java
package org.example;

import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class Game {

    private Set<Cell> aliveCells;

    public Game(Set<Cell> aliveCells) {
        this.aliveCells = aliveCells;
    }

    public Set<Cell> getAliveCells() {
        return aliveCells;
    }

    public void start(int numberOfIterations) {
        for (int i = 0; i < numberOfIterations; i++) {
            nextIteration();
        }
    }

    private void nextIteration() {
        Set<Cell> cellsToReborn = aliveCells.stream()
                .flatMap(cell -> cell.calculatePossibleNeighbours().stream())
                .distinct()
                .filter(cell -> !aliveCells.contains(cell))
                .filter(cell -> calculateAliveNeighboursCount(cell) == 3)
                .collect(Collectors.toSet());


        Set<Cell> newAliveCells = aliveCells.stream()
                .filter(cell -> {
                    int count = calculateAliveNeighboursCount(cell);
                    return count == 2 || count == 3;
                })
                .collect(Collectors.toSet());

        aliveCells = Stream.concat(cellsToReborn.stream(), newAliveCells.stream()).collect(Collectors.toSet());
    }

    private int calculateAliveNeighboursCount(Cell cell) {
        Set<Cell> possibleNeighbours = cell.calculatePossibleNeighbours();
        possibleNeighbours.retainAll(aliveCells);
        return possibleNeighbours.size();
    }


    public boolean isCellAlive(Cell cell) {
        return aliveCells.contains(cell);
    }
}

File: kotlin/src/main/kotlin/Game.kt
import java.util.stream.Collectors
import java.util.stream.Stream

internal class Game(private var aliveCells: Set<Cell>) {

    fun getAliveCells(): Set<Cell> {
        return aliveCells
    }

    fun start(numberOfIterations: Int) {
        for (i in 0..<numberOfIterations) {
            nextIteration()
        }
    }

    private fun nextIteration() {
        val cellsToReborn: Set<Cell> = aliveCells.stream()
            .flatMap { cell: Cell ->
                cell.calculatePossibleNeighbours().stream()
            }
            .distinct()
            .filter { cell: Cell -> !aliveCells.contains(cell) }
            .filter { cell: Cell -> calculateAliveNeighboursCount(cell) == 3 }
            .collect(Collectors.toSet())

        val newAliveCells: Set<Cell> = aliveCells.stream()
            .filter { cell: Cell ->
                val count = calculateAliveNeighboursCount(cell)
                count == 2 || count == 3
            }
            .collect(Collectors.toSet())
        
        aliveCells = Stream.concat(cellsToReborn.stream(), newAliveCells.stream())
            .collect(Collectors.toSet())
    }

    private fun calculateAliveNeighboursCount(cell: Cell): Int {
        val possibleNeighbours: MutableSet<Cell> = cell.calculatePossibleNeighbours().toMutableSet()
        possibleNeighbours.retainAll(aliveCells)
        return possibleNeighbours.size
    }

    fun isCellAlive(cell: Cell): Boolean {
        return aliveCells.contains(cell)
    }


    fun main(args: Array<String>) {}
}
